game-core/src/test/java/games/strategy/triplea/ai/pro/util/MoveBatcherTest.java

```diff
@@ -0,0 +1,130 @@
+
+import com.google.common.collect.Maps;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.xml.TestMapGameData;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+public class MoveBatcherTest {
+  private final GameData gameData = TestMapGameData.REVISED.getGameData();
+  private final Territory brazil = territory("Brazil", gameData);
+  private final Territory sz19 = territory("19 Sea Zone", gameData);
+  private final Territory sz18 = territory("18 Sea Zone", gameData);
+  private final Territory sz12 = territory("12 Sea Zone", gameData);
+  private final Territory algeria = territory("Algeria", gameData);
+  private final Route brazilToSz18 = new Route(brazil, sz18);
+  private final Route sz19ToSz18 = new Route(sz19, sz18);
+  private final Route sz18ToSz12 = new Route(sz18, sz12);
+  private final Route sz12ToAlgeria = new Route(sz12, algeria);
+  private final Route algeriaToSz18 = new Route(sz12, algeria);
+  private final Unit inf1 = infantry(gameData).create(americans(gameData));
+  private final Unit inf2 = infantry(gameData).create(americans(gameData));
+  private final Unit tank1 = armour(gameData).create(americans(gameData));
+  private final Unit tank2 = armour(gameData).create(americans(gameData));
+  private final Unit transport1 = transport(gameData).create(americans(gameData));
+  private final Unit transport2 = transport(gameData).create(americans(gameData));
+
+  public MoveBatcherTest() throws Exception {}
+
+  private static ArrayList<Unit> unitList(final Unit... units) {
+    return new ArrayList<Unit>(List.of(units));
+  }
+
+  @Test
+  public void testMoveUnitsWithMultipleTransports() {
+    final MoveBatcher moves = new MoveBatcher();
+
+    moves.newSequence();
+    moves.addTransportLoad(inf1, brazilToSz18, transport1);
+    moves.addTransportLoad(tank1, brazilToSz18, transport1);
+    moves.addMove(unitList(transport1, tank1, inf1), sz18ToSz12);
+    moves.addMove(unitList(tank1, inf1), sz12ToAlgeria);
+
+    moves.newSequence();
+    moves.addTransportLoad(tank2, brazilToSz18, transport2);
+    moves.addTransportLoad(inf2, brazilToSz18, transport2);
+    moves.addMove(unitList(transport2, inf2, tank2), sz18ToSz12);
+    moves.addMove(unitList(inf2, tank2), sz12ToAlgeria);
+
+    final List<Collection<Unit>> moveUnits = new ArrayList<>();
+    final List<Route> moveRoutes = new ArrayList<>();
+    final List<Map<Unit, Unit>> unitsToTransports = new ArrayList<>();
+    moves.batchAndEmit(moveUnits, moveRoutes, unitsToTransports);
+
+    assertEquals(3, moveUnits.size());
+    assertEquals(3, moveRoutes.size());
+    assertEquals(3, unitsToTransports.size());
+
+    assertEquals(unitList(inf1, tank1, tank2, inf2), moveUnits.get(0));
```

Comment:
If this assertion were to fail, would it be clear to someone why it was expected to pass in the first place? Could that be answered with minimal effort? If a change were made, or a series of changes made, and all assertions in this test were to fail, how much time would it take for someone to determine why the assertions were valid in the first place and hence the necessary modification?

To reduce this cost, as a first pass, I recommend adding a comment to each assertion, for example `assertEquals("these units were unloaded together, should be grouped together", unitLIst(...), moveUnits.get(0))`